cmd/remote_val_harness/internal/test_harness.go

```diff
@@ -0,0 +1,363 @@
+		return nil, newTestHarnessError(ErrFailedToExpandPath, err, cfg.StateFile)
+	}
+	logger.Info("Loading private validator configuration", "keyFile", keyFile, "stateFile", stateFile)
+	// NOTE: LoadFilePV ultimately calls os.Exit on failure. No error will be
+	// returned if this call fails.
+	fpv := privval.LoadFilePV(keyFile, stateFile)
+
+	if genesisFile, err = expandPath(cfg.GenesisFile); err != nil {
+		return nil, newTestHarnessError(ErrFailedToExpandPath, err, cfg.GenesisFile)
+	}
+	logger.Info("Loading chain ID from genesis file", "genesisFile", genesisFile)
+	st, err := state.MakeGenesisDocFromFile(genesisFile)
+	if err != nil {
+		return nil, newTestHarnessError(ErrFailedToLoadGenesisFile, err, genesisFile)
+	}
+	logger.Info("Loaded genesis file", "chainID", st.ChainID)
+
+	sc, err := newTestHarnessSocketVal(logger, cfg)
+	if err != nil {
+		return nil, newTestHarnessError(ErrFailedToCreateListener, err, "")
+	}
+
+	return &TestHarness{
+		sc:      sc,
+		fpv:     fpv,
+		chainID: st.ChainID,
+		logger:  logger,
+	}, nil
+}
+
+func (th *TestHarness) Run() {
+	th.logger.Info("Starting test harness")
+	donec := make(chan struct{})
+	go func() {
+		defer close(donec)
+		if err := th.sc.Start(); err != nil {
+			th.logger.Error("Failed to start listener", "err", err)
+			th.Shutdown(newTestHarnessError(ErrFailedToStartListener, err, ""))
+		}
+		if err := th.TestPublicKey(); err != nil {
+			th.Shutdown(err)
+		}
+		if err := th.TestSignProposal(); err != nil {
+			th.Shutdown(err)
+		}
+		if err := th.TestSignVote(); err != nil {
+			th.Shutdown(err)
+		}
+		th.logger.Info("SUCCESS! All tests passed.")
+	}()
+
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+	go func() {
+		for sig := range c {
+			th.logger.Info("Caught interrupt, terminating...", "sig", sig)
+			th.Shutdown(newTestHarnessError(ErrInterrupted, nil, ""))
+		}
+	}()
+
+	// Run until complete
+	<-donec
+	th.Shutdown(nil)
+}
+
+func (th *TestHarness) TestPublicKey() error {
+	th.logger.Info("TEST: Public key of remote signer")
+	th.logger.Info("Local", "pubKey", th.fpv.GetPubKey())
+	th.logger.Info("Remote", "pubKey", th.sc.GetPubKey())
+	if th.fpv.GetPubKey() != th.sc.GetPubKey() {
+		th.logger.Error("FAILED: Local and remote public keys do not match")
+		return newTestHarnessError(ErrTestPublicKeyFailed, nil, "")
+	}
+	return nil
+}
+
+func (th *TestHarness) TestSignProposal() error {
+	th.logger.Info("TEST: Signing of proposals")
+	// sha256 hash of "hash"
+	hash, _ := hex.DecodeString("D04B98F48E8F8BCC15C6AE5AC050801CD6DCFD428FB5F9E65C4E16E7807340FA")
+	prop := &types.Proposal{
+		Type:     types.ProposalType,
+		Height:   12345,
+		Round:    23456,
+		POLRound: -1,
+		BlockID: types.BlockID{
+			Hash: hash,
+			PartsHeader: types.PartSetHeader{
+				Hash:  hash,
+				Total: 1000000,
+			},
+		},
+		Timestamp: time.Now(),
+	}
+	propBytes, err := cdc.MarshalBinaryLengthPrefixed(types.CanonicalizeProposal(th.chainID, prop))
```

Comment:
Mostly because of the panic over [here](https://github.com/tendermint/tendermint/blob/develop/types/proposal.go#L94). I'd prefer to return any error and try to shut down gracefully, ideally.

Reply:
Thanks for the explanation. This should only panic if we completely messed up either amino or introduced sth obscure in `CanonicalizeProposal` that can't be encoded. IMHO, I still think this isn't sth. that this integration tests should bother about.

Reply:
Okay cool, I'll use the `SignBytes` function. I suppose, like you say, if it does panic it's a sign of something pretty important that needs to be fixed ASAP 😁